File: app/src/main/java/com/wirelessalien/android/moviedb/fragment/LoginFragment.java
/*
 *     This file is part of Movie DB. <https://github.com/WirelessAlien/MovieDB>
 *     forked from <https://notabug.org/nvb/MovieDB>
 *
 *     Copyright (C) 2024  WirelessAlien <https://github.com/WirelessAlien>
 *
 *     Movie DB is free software: you can redistribute it and/or modify
 *     it under the terms of the GNU General Public License as published by
 *     the Free Software Foundation, either version 3 of the License, or
 *     (at your option) any later version.
 *
 *     Movie DB is distributed in the hope that it will be useful,
 *     but WITHOUT ANY WARRANTY; without even the implied warranty of
 *     MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *     GNU General Public License for more details.
 *
 *     You should have received a copy of the GNU General Public License
 *     along with Movie DB.  If not, see <https://www.gnu.org/licenses/>.
 */

package com.wirelessalien.android.moviedb.fragment;

import android.content.ComponentName;
import android.content.Intent;
import android.content.SharedPreferences;
import android.net.Uri;
import android.os.Bundle;
import android.os.Handler;
import android.os.Looper;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.Button;
import android.widget.TextView;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.browser.customtabs.CustomTabsClient;
import androidx.browser.customtabs.CustomTabsIntent;
import androidx.browser.customtabs.CustomTabsServiceConnection;
import androidx.preference.PreferenceManager;

import com.google.android.material.bottomsheet.BottomSheetDialogFragment;
import com.squareup.picasso.Picasso;
import com.wirelessalien.android.moviedb.R;
import com.wirelessalien.android.moviedb.tmdb.account.GetAccountDetailsThread;
import com.wirelessalien.android.moviedb.tmdb.account.LogoutThread;
import com.wirelessalien.android.moviedb.tmdb.account.TMDbAuthThreadV4;

import de.hdodenhof.circleimageview.CircleImageView;

public class LoginFragment extends BottomSheetDialogFragment {

    private SharedPreferences preferences;

    @Override
    public void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        preferences = PreferenceManager.getDefaultSharedPreferences(requireContext());
    }

    @Nullable
    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, @Nullable ViewGroup container, @Nullable Bundle savedInstanceState) {
        View view = inflater.inflate(R.layout.fragment_login, container, false);

        preferences = PreferenceManager.getDefaultSharedPreferences(requireContext());

        CustomTabsIntent customTabsIntent = new CustomTabsIntent.Builder().build();
        customTabsIntent.intent.setPackage("com.android.chrome");
        CustomTabsClient.bindCustomTabsService(requireContext(), "com.android.chrome", new CustomTabsServiceConnection() {
            @Override
            public void onCustomTabsServiceConnected(@NonNull ComponentName componentName, @NonNull CustomTabsClient customTabsClient) {
                customTabsClient.warmup(0L);
            }

            @Override
            public void onServiceDisconnected(ComponentName name) {}
        });

        Button loginButton = view.findViewById(R.id.login);
        Button logoutButton = view.findViewById(R.id.logout);
        TextView loginStatus = view.findViewById(R.id.login_status);
        Button signUpButton = view.findViewById(R.id.signup);
        TextView nameTextView = view.findViewById(R.id.name);
        CircleImageView avatar = view.findViewById(R.id.avatar);

        signUpButton.setOnClickListener(v -> {
            String url = "https://www.themoviedb.org/signup";
            Intent i = new Intent( Intent.ACTION_VIEW);
            i.setData( Uri.parse(url));
            startActivity(i);
        });


        if (preferences.getString("account_id", null) != null && preferences.getString("access_token", null) != null){
            loginButton.setVisibility(View.GONE);
            logoutButton.setVisibility(View.VISIBLE);
            loginStatus.setText("You are logged in");

        }

        logoutButton.setOnClickListener(v -> {
            Handler handler = new Handler( Looper.getMainLooper());
            LogoutThread logoutThread = new LogoutThread(requireContext(), handler);
            logoutThread.start();

            loginButton.setVisibility(View.VISIBLE);
            logoutButton.setVisibility(View.GONE);
            loginStatus.setText("You are not logged in");
        });

        loginButton.setOnClickListener(v -> {

            TMDbAuthThreadV4 authTask = new TMDbAuthThreadV4(getContext()) {
                @Override
                public void run() {
                    String accessToken = authenticate();
                    if (accessToken != null) {
                        preferences.edit().putString("request_token", accessToken).apply();
                    }
                }
            };
            authTask.start();
        });

        //if access token is not null, then the user is logged in
        if (preferences.getString("access_token", null) != null){
            GetAccountDetailsThread getAccountIdThread = new GetAccountDetailsThread(getContext(), (accountId, name, username, avatarPath, gravatar) -> requireActivity().runOnUiThread( () -> {
                // If both username and name are available, display name. Otherwise, display whichever is available.
                if (name != null && !name.isEmpty()) {
                    nameTextView.setText(name);
                } else if (username != null && !username.isEmpty()) {
                    nameTextView.setText(username);
                }

                // If both avatarPath and gravatar are available, display avatarPath. Otherwise, display whichever is available.
                if (avatarPath != null && !avatarPath.isEmpty()) {
                    Picasso.get().load("https://image.tmdb.org/t/p/w200" + avatarPath).into(avatar);
                } else if (gravatar != null && !gravatar.isEmpty()) {
                    Picasso.get().load("https://secure.gravatar.com/avatar/" + gravatar + ".jpg?s=200").into(avatar);
                }
            } ) );
            getAccountIdThread.start();
        }

        return view;
    }
}

File: app/src/main/java/com/wirelessalien/android/moviedb/fragment/LoginFragment.kt
/*
 *     This file is part of "ShowCase" formerly Movie DB. <https://github.com/WirelessAlien/MovieDB>
 *     forked from <https://notabug.org/nvb/MovieDB>
 *
 *     Copyright (C) 2024  WirelessAlien <https://github.com/WirelessAlien>
 *
 *     ShowCase is free software: you can redistribute it and/or modify
 *     it under the terms of the GNU General Public License as published by
 *     the Free Software Foundation, either version 3 of the License, or
 *     (at your option) any later version.
 *
 *     ShowCase is distributed in the hope that it will be useful,
 *     but WITHOUT ANY WARRANTY; without even the implied warranty of
 *     MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *     GNU General Public License for more details.
 *
 *     You should have received a copy of the GNU General Public License
 *     along with "ShowCase".  If not, see <https://www.gnu.org/licenses/>.
 */
package com.wirelessalien.android.moviedb.fragment

import android.content.ComponentName
import android.content.Intent
import android.content.SharedPreferences
import android.net.Uri
import android.os.Bundle
import android.os.Handler
import android.os.Looper
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.browser.customtabs.CustomTabsClient
import androidx.browser.customtabs.CustomTabsIntent
import androidx.browser.customtabs.CustomTabsServiceConnection
import androidx.lifecycle.lifecycleScope
import androidx.preference.PreferenceManager
import com.google.android.material.bottomsheet.BottomSheetDialogFragment
import com.squareup.picasso.Picasso
import com.wirelessalien.android.moviedb.R
import com.wirelessalien.android.moviedb.databinding.FragmentLoginBinding
import com.wirelessalien.android.moviedb.helper.ConfigHelper
import com.wirelessalien.android.moviedb.tmdb.account.AccountLogout
import com.wirelessalien.android.moviedb.tmdb.account.GetAccountDetails
import com.wirelessalien.android.moviedb.tmdb.account.TMDbAuthV4
import kotlinx.coroutines.launch

class LoginFragment : BottomSheetDialogFragment() {
    private lateinit var preferences: SharedPreferences
    private var clientId: String? = null
    private lateinit var binding: FragmentLoginBinding

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        preferences = PreferenceManager.getDefaultSharedPreferences(requireContext())
    }

    override fun onCreateView(
        inflater: LayoutInflater,
        container: ViewGroup?,
        savedInstanceState: Bundle?
    ): View {
        binding = FragmentLoginBinding.inflate(inflater, container, false)
        val view = binding.root
        preferences = PreferenceManager.getDefaultSharedPreferences(requireContext())
        clientId = ConfigHelper.getConfigValue(requireContext().applicationContext, "client_id")

        val customTabsIntent = CustomTabsIntent.Builder().build()
        customTabsIntent.intent.setPackage("com.android.chrome")
        CustomTabsClient.bindCustomTabsService(
            requireContext(),
            "com.android.chrome",
            object : CustomTabsServiceConnection() {
                override fun onCustomTabsServiceConnected(
                    componentName: ComponentName,
                    customTabsClient: CustomTabsClient
                ) {
                    customTabsClient.warmup(0L)
                }

                override fun onServiceDisconnected(name: ComponentName) {}
            })

        binding.signup.setOnClickListener {
            val url = "https://www.themoviedb.org/signup"
            val i = Intent(Intent.ACTION_VIEW)
            i.setData(Uri.parse(url))
            startActivity(i)
        }

        if (preferences.getString("account_id", null) != null && preferences.getString("access_token", null) != null) {
            binding.login.visibility = View.GONE
            binding.logout.visibility = View.VISIBLE
            binding.loginStatus.setText(R.string.logged_in)
        }

        binding.logout.setOnClickListener {
            lifecycleScope.launch {
                val logoutManager = AccountLogout(requireContext(), Handler(Looper.getMainLooper()))
                logoutManager.logout()
                binding.login.visibility = View.VISIBLE
                binding.logout.visibility = View.GONE
                binding.loginStatus.setText(R.string.not_logged_in)
                binding.name.visibility = View.GONE
            }
        }

        binding.login.setOnClickListener {
            lifecycleScope.launch {
                val authCoroutine = TMDbAuthV4(requireContext())
                val accessToken = authCoroutine.authenticate()
                if (accessToken != null) {
                    preferences.edit().putString("access_token", accessToken).apply()
                }
            }
        }

        binding.changeProvider.setOnClickListener {
            dismiss()
            val bottomSheetFragment = LoginFragmentTkt()
            bottomSheetFragment.show(parentFragmentManager, bottomSheetFragment.tag)

        }

        if (preferences.getString("access_token", null) != null) {
            lifecycleScope.launch {
                val getAccountDetailsCoroutine = GetAccountDetails(requireContext(), object : GetAccountDetails.AccountDataCallback {
                    override fun onAccountDataReceived(accountId: Int, name: String?, username: String?, avatarPath: String?, gravatar: String?) {
                        if (isAdded) {
                            requireActivity().runOnUiThread {
                                binding.name.visibility = View.VISIBLE
                                if (!name.isNullOrEmpty()) {
                                    binding.name.text = name
                                } else if (!username.isNullOrEmpty()) {
                                    binding.name.text = username
                                }

                                if (!avatarPath.isNullOrEmpty()) {
                                    Picasso.get().load("https://image.tmdb.org/t/p/w200$avatarPath").into(binding.avatar)
                                } else if (!gravatar.isNullOrEmpty()) {
                                    Picasso.get().load("https://secure.gravatar.com/avatar/$gravatar.jpg?s=200").into(binding.avatar)
                                }
                            }
                        }
                    }
                })
                getAccountDetailsCoroutine.fetchAccountDetails()
            }
        }
        return view
    }

    private fun redirectToTraktAuthorization() {
        val redirectUri = "trakt.wirelessalien.showcase://callback"
        val authUrl = "https://trakt.tv/oauth/authorize?response_type=code&client_id=$clientId&redirect_uri=$redirectUri"

        val intent = Intent(Intent.ACTION_VIEW, Uri.parse(authUrl))
        startActivity(intent)
    }
}